Service/ExecuteSourceDeductionForItems.php

```diff
@@ -0,0 +1,148 @@
+     */
+    private $salesEventFactory;
+
+    /**
+     * @var ItemToDeductFactory
+     */
+    private $itemToDeductFactory;
+
+    /**
+     * @var SourceDeductionRequestFactory
+     */
+    private $sourceDeductionRequestFactory;
+
+    /**
+     * @var SourceDeductionService
+     */
+    private $sourceDeductionService;
+
+    /**
+     * @var SourcesRepositoryInterface
+     */
+    private $sourceRepository;
+
+    /**
+     * ExecuteSourceDeductionForItems constructor.
+     * @param WebsiteRepositoryInterface $websiteRepository
+     * @param SalesEventInterfaceFactory $salesEventFactory
+     * @param SalesChannelInterfaceFactory $salesChannelFactory
+     * @param ItemToDeductFactory $itemToDeductFactory
+     * @param SourceDeductionRequestFactory $sourceDeductionRequestFactory
+     * @param SourceDeductionService $sourceDeductionService
+     * @param SourcesRepositoryInterface $sourceRepository
+     * @param Processor $priceIndexer
+     */
+    public function __construct(
+        WebsiteRepositoryInterface $websiteRepository,
+        SalesEventInterfaceFactory $salesEventFactory,
+        SalesChannelInterfaceFactory $salesChannelFactory,
+        ItemToDeductFactory $itemToDeductFactory,
+        SourceDeductionRequestFactory $sourceDeductionRequestFactory,
+        SourceDeductionService $sourceDeductionService,
+        SourcesRepositoryInterface $sourceRepository,
+        Processor $priceIndexer
+    ) {
+        $this->websiteRepository = $websiteRepository;
+        $this->salesEventFactory = $salesEventFactory;
+        $this->salesChannelFactory = $salesChannelFactory;
+        $this->itemToDeductFactory = $itemToDeductFactory;
+        $this->sourceDeductionRequestFactory = $sourceDeductionRequestFactory;
+        $this->sourceDeductionService = $sourceDeductionService;
+        $this->sourceRepository = $sourceRepository;
+        $this->priceIndexer = $priceIndexer;
+    }
+
+    /**
+     * @param OrderInterface $order
+     * @param GetItemsToCancelFromOrderItem[] $itemsToCancel
+     * @param Creditmemo|null $creditmemo
+     */
+    public function executeSourceDeductionForItems(OrderInterface $order, array $itemsToCancel, Creditmemo $creditmemo = null)
+    {
+        $type = SalesEventInterface::EVENT_ORDER_CANCELED;
+        if ($creditmemo) {
+            $type = SalesEventInterface::EVENT_CREDITMEMO_CREATED;
+        }
+
+        $websiteId = $order->getStore()->getWebsiteId();
+        $websiteCode = $this->websiteRepository->getById($websiteId)->getCode();
+        $salesChannel = $this->salesChannelFactory->create([
+            'data' => [
+                'type' => SalesChannelInterface::TYPE_WEBSITE,
+                'code' => $websiteCode
+            ]
+        ]);
+
+        $salesEvent = $this->salesEventFactory->create([
+            'type' => $type,
+            'objectType' => SalesEventInterface::OBJECT_TYPE_ORDER,
+            'objectId' => (string)$order->getEntityId()
```

Comment:
`getEntityId`, we should probably be using `getId`, I think I've mentioned before about how this method will resolve the correct unique/auto-increment ID for a row in the DB